python/pyiceberg/io/s3fs.py

```diff
@@ -0,0 +1,235 @@
+
+        Raises:
+            FileExistsError: If the file already exists at `self.location` and `overwrite` is False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        self._s3.touch(self.location)
+        return S3fsOutputStream(s3_object=self._s3.open(self.location, "wb"))
+
+    def to_input_file(self) -> S3fsInputFile:
+        """Returns a new S3fsInputFile for the location at `self.location`"""
+        return S3fsInputFile(location=self.location, s3=self._s3)
+
+
+class S3fsFileIO(FileIO):
+    """A FileIO implementation that uses s3fs"""
+
+    def __init__(self, client_kwargs: Optional[dict] = None, **s3fs_filesystem_kwargs):
+        if client_kwargs:
+            self._s3_filesystem = S3FileSystem(client_kwargs=client_kwargs, **s3fs_filesystem_kwargs)
+        else:
+            self._s3_filesystem = S3FileSystem(**s3fs_filesystem_kwargs)
+
+    def new_input(self, location: str) -> S3fsInputFile:
+        """Get an S3fsInputFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            S3fsInputFile: An S3fsInputFile instance for the given location
+        """
+        return S3fsInputFile(location=location, s3=self._s3_filesystem)
+
+    def new_output(self, location: str) -> S3fsOutputFile:
+        """Get an S3fsOutputFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            S3fsOutputFile: An S3fsOutputFile instance for the given location
+        """
+        return S3fsOutputFile(location=location, s3=self._s3_filesystem)
+
+    def delete(self, location: Union[str, InputFile, OutputFile]) -> None:
+        """Delete the file at the given location
+
+        Args:
+            location(str, InputFile, OutputFile): The URI to the file--if an InputFile instance or an
+            OutputFile instance is provided, the location attribute for that instance is used as the location
+            to delete
+        """
+        # Create a ParseResult from the uri
+        if isinstance(location, (InputFile, OutputFile)):
+            str_location = location.location  # Use InputFile or OutputFile location
+        else:
+            str_location = location
+
+        self._s3_filesystem.rm(str_location)
```

Comment:
Should we explicitly pass `recursive=False`? I know it's the default but it doesn't seem like something we want to mess up if the underlying library changes.